popde/density_estimate.py

```diff
@@ -93,6 +93,67 @@ def evaluate_scipy(self, points):
         return density_values
 
 
+    def plot_2d_contour(self, dim1, dim2, slice_dims=None, slice_values=None, num_points=100, **kwargs):
+        """
+        Plot a 2D contour of the KDE with optional slicing along other dimensions.
+
+        Parameters:
+            - dim1, dim2: Dimensions for the plot axes.
+            - slice_dims: Dimensions to slice along (list or tuple).
+            - slice_values: Values for slicing along slice_dims (list or tuple).
+            - num_points: Number of points for the contour plot.
+            - **kwargs: Additional keyword arguments passed to the `contour` function.
+        Example:
+            np.random.seed(42)
+            # Number of data points
+            num_points = 1000
+            # Mean and covariance matrix
+            mean = [0, 0, 0]
+            covariance_matrix = [[1, 0.5, 0.3],
+                                [0.5, 1, 0.2],
+                                [0.3, 0.2, 1]]
+
+            # Generate 3D normal distributed data
+            data = np.random.multivariate_normal(mean, covariance_matrix, num_points)
+            parameter = ['m1', 'm2', 'Mc']
+            # Extract individual variables
+            x, y, z = data[:, 0], data[:, 1], data[:, 2]
+            kde = (data)
+
+            # Plot a 2D contour with a slice along the 'z' dimension
+            kde.plot_2d_contour(parameter[0], parameter[1], slice_dims=[parameter[2]], slice_values=[0], num_points=100, colors='blue', alpha=0.5)
+
+        """
+        # Generate a grid for the contour plot
+        dim1_min, dim1_max = np.min(self.data[dim1]), np.max(self.data[dim1])
+        dim2_min, dim2_max = np.min(self.data[dim2]), np.max(self.data[dim2])
+        dim1_grid = np.linspace(dim1_min, dim1_max, num_points)
+        dim2_grid = np.linspace(dim2_min, dim2_max, num_points)
+        xx, yy = np.meshgrid(dim1_grid, dim2_grid)
+        positions = np.column_stack([xx.ravel(), yy.ravel()])
+
+        # If slicing is specified, insert the slice values into the positions array
+        if slice_dims is not None and slice_values is not None:
+            for slice_dim, slice_value in zip(slice_dims, slice_values):
+                slice_idx = list(self.data.columns).index(slice_dim)
+                positions = np.insert(positions, slice_idx, slice_value, axis=0)
+
+        # Evaluate the KDE at the grid points need fix here
+        z = self.kde(positions)
+
+        # Reshape the results for contour plotting
+        zz = z.reshape(xx.shape)
+
+        # Create the contour plot
+        plt.contour(xx, yy, zz, **kwargs)
+        plt.xlabel(dim1)
+        plt.ylabel(dim2)
+        plt.title(f'2D Contour Plot of KDE for {dim1} and {dim2} (Sliced along {slice_dims})')
+        plt.savefig("KDE"+dim1+dim2+"_contour.png")
+        plt.show()
```

Comment:
as I said for the plot utils function you should not assume this is being run on a system where you can just `show` a plot, this will not work on a cluster.  Bear in mind that this method will be called inside a script so you can do `return fig` here (after creating a fig object) to allow someone to do `show` inside the script if wanted. 

Also there should be the option of _not_ saving the plot, eg if the method is called with file name = None

Reply:
done